mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp

```diff
@@ -412,163 +402,183 @@ struct BlockwiseGemmV2RewritePattern
+      Value kBaseB = ilkb.create<AddIOp>(loc, bBase, blk_td);
 
-      // for(index_t k_i = 0; k_i < KPerThread; k_i += mfma_type.num_input_blks)
-      // {
-      //     a[k_i] = p_a_wave[(k_i * num_input_blks + blk_id) * M + blk_td];
-      //     b[k_i] = p_b_wave[(k_i * num_input_blks + blk_id) * N + blk_td];
+      // for(index_t k_i = 0; k_i < KPerThread; k_i +=
+      // mfma_type.num_input_blks) {
+      //   a[k_i] = p_a_wave[(k_i * num_input_blks + blk_id) * M + blk_td];
+      //   b[k_i] = p_b_wave[(k_i * num_input_blks + blk_id) * N + blk_td];
       // }
       // p_a_wave need to be offseted by waveOffsetA.
       // p_b_wave need to be offseted by waveOffsetB.
 
       auto inputSpansPerMfmaInConstantOp =
-          b.create<ConstantIndexOp>(loc, inputSpansPerMfmaIn);
-
-      auto loopKLoad = b.create<AffineForOp>(loc, 0, KPerThread);
-      auto lklb = ConversionPatternRewriter::atBlockBegin(loopKLoad.getBody(),
-                                                          b.getListener());
-      auto lkliv = loopKLoad.getInductionVar();
+          ilkb.create<ConstantIndexOp>(loc, inputSpansPerMfmaIn);
 
-      Value sourceOffsetA = lklb.create<AddIOp>(
+      Value sourceOffsetA = ilkb.create<AddIOp>(
           loc,
-          lklb.create<MulIOp>(
+          ilkb.create<MulIOp>(
               loc,
-              lklb.create<AddIOp>(
+              ilkb.create<AddIOp>(
                   loc,
-                  lklb.create<MulIOp>(loc, lkliv,
+                  ilkb.create<MulIOp>(loc, ilkiv,
                                       inputSpansPerMfmaInConstantOp),
                   blk_id),
               MConstantOp),
           kBaseA);
-
-      if (KPack > 1)
-        sourceOffsetA = lklb.create<MulIOp>(
-            loc, sourceOffsetA, lklb.create<ConstantIndexOp>(loc, KPack));
-
-      Value valueA = lklb.create<InBoundsLoadOp>(
-          loc, bufferAElementType, op.getMatrixA(), sourceOffsetA);
-      lklb.create<memref::StoreOp>(loc, valueA, bufferA, ValueRange{lkliv});
-
-      Value sourceOffsetB = lklb.create<AddIOp>(
+      Value sourceOffsetB = ilkb.create<AddIOp>(
           loc,
-          lklb.create<MulIOp>(
+          ilkb.create<MulIOp>(
               loc,
-              lklb.create<AddIOp>(
+              ilkb.create<AddIOp>(
                   loc,
-                  lklb.create<MulIOp>(loc, lkliv,
+                  ilkb.create<MulIOp>(loc, ilkiv,
                                       inputSpansPerMfmaInConstantOp),
                   blk_id),
               NConstantOp),
           kBaseB);
 
-      if (KPack > 1)
-        sourceOffsetB = lklb.create<MulIOp>(
-            loc, sourceOffsetB, lklb.create<ConstantIndexOp>(loc, KPack));
-
-      Value valueB = lklb.create<InBoundsLoadOp>(
-          loc, bufferBElementType, op.getMatrixB(), sourceOffsetB);
-      lklb.create<memref::StoreOp>(loc, valueB, bufferB, ValueRange{lkliv});
+      if (KPack > 1) {
```

Comment:
Am I missing something, or is this code basically repeated twice?

Reply:
It somewhat repeated twice between the kpack > 1 and kpack = 1 conditions because we have to do differently between the two types. I was reminded about this in the overdue ticket that uniform kpack into a dimension as well. But this PR is geared towards performance (and evaluation). I'd very well like to decouple the refactoring ticket to a different PR.

Reply:
I'd like to use the fact that we're doing a rewrite for performance as a reason to also do the refactoring so we can't keep putting off the refactoring for forever.

Reply:
Right at the moment I don't seek for merging this PR until I got clear experiment feedback on performance improvements. I think I agree with you in general, that refactoring is good, but the question is about when.

As you can see in the past experience, the refactoring PRs (transform-maps in threadwisegemm) that went through ended up getting discarded because we'd no longer need a coordinate transform in the inner k loop. I'd argue that putting off the pure refactoring PRs has a benefit (in my efficiency) when we fast prototype.

Talking about this PR, I may end up discarding it if experiment shows a performance degrade. If there's a perf improvement, and we decide to keep it, I'd still leaning towards doing refactoring in a follow-up PR, just not to pack all changes together. We should create ticket and assign to next milestone if you have concerns on the when problem.